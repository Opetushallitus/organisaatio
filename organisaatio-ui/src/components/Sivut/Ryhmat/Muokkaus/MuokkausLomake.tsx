import * as React from 'react';
import styles from './MuokkausLomake.module.css';
import Icon from '@iconify/react';
import homeIcon from '@iconify/icons-fa-solid/home';
import Input from '@opetushallitus/virkailija-ui-components/Input';
import Select from '@opetushallitus/virkailija-ui-components/Select';
import { Ryhma } from '../../../../types/types';
import Button from '@opetushallitus/virkailija-ui-components/Button';
import PohjaSivu from '../../PohjaSivu/PohjaSivu';
import { useContext } from 'react';
import { KoodistoContext, LanguageContext } from '../../../../contexts/contexts';
import { mapLocalizedKoodiToLang } from '../../../../tools/mappers';
import { FieldValues } from 'react-hook-form/dist/types/fields';
import { SubmitHandler, useForm, Controller } from 'react-hook-form';
import { joiResolver } from '@hookform/resolvers/joi';
import { Link } from 'react-router-dom';
import RyhmatLomakeSchema from '../../../../ValidationSchemas/RyhmatLomakeSchema';

export type MuokkausLomakeProps = {
    onUusi: boolean;
    ryhma: Ryhma;
    handlePeruuta: () => void;
    handlePassivoi: () => void;
    handlePoista: () => void;
    handleTallenna: SubmitHandler<FieldValues>;
};

const MuokkausLomake = ({
    onUusi,
    ryhma,
    handlePeruuta,
    handlePassivoi,
    handlePoista,
    handleTallenna,
}: MuokkausLomakeProps) => {
    const { i18n, language } = useContext(LanguageContext);
    const { ryhmaTyypitKoodisto, kayttoRyhmatKoodisto } = useContext(KoodistoContext);

    const ryhmaTyypitOptions = ryhmaTyypitKoodisto.selectOptions();
    const kayttoRyhmatOptions = kayttoRyhmatKoodisto.selectOptions();

    const kayttoRyhmat = ryhma.kayttoryhmat.map((koodiUri) => kayttoRyhmatKoodisto.uri2SelectOption(koodiUri)); //mapValuesToSelect(ryhma.kayttoryhmat, kayttoRyhmatOptions);
    const ryhmaTyypit = ryhma.ryhmatyypit.map((koodiUri) => ryhmaTyypitKoodisto.uri2SelectOption(koodiUri)); //mapValuesToSelect(ryhma.ryhmatyypit, ryhmaTyypitOptions);
    const isDisabled = !ryhma || ryhma.status === 'PASSIIVINEN';
    const {
        register,
        formState: { errors: validationErrors },
        handleSubmit,
        control,
    } = useForm({ resolver: joiResolver(RyhmatLomakeSchema) });

    return (
        <PohjaSivu>
            <div className={styles.YlaBanneri}>
                <div>
                    <Link to="/ryhmat">
                        <Icon icon={homeIcon} />
                    </Link>
                </div>
                <div>
                    <Link to="/ryhmat">{i18n.translate('RYHMAT_KAIKKI_RYHMAT')}</Link>
                </div>
            </div>
            <div className={styles.PaaKehys}>
                <div className={styles.ValiContainer}>
                    <div className={styles.ValiOtsikko}>
                        <h3>{i18n.translate('RYHMAT_RYHMA')}</h3>
                        <h1>{mapLocalizedKoodiToLang(language, 'nimi', ryhma)}</h1>
                    </div>
                </div>
                <div className={styles.PaaOsio}>
                    <div className={styles.OtsikkoRivi}>
                        <div className={styles.Otsikko}>
                            <h3>{i18n.translate('RYHMAT_RYHMAN_TIEDOT_OTSIKKO')}</h3>
                        </div>
                    </div>
                    <div className={styles.OidRivi}>
                        <span className={styles.AvainKevyestiBoldattu}>{i18n.translate('LABEL_OID')}</span>
                        <span className={styles.ReadOnly}>{ryhma.oid || ''}</span>
                    </div>
                    <div className={styles.Rivi}>
                        <div className={styles.Kentta}>
                            <label>{i18n.translate('RYHMAT_RYHMAN_NIMI')}</label>
                            <div className={styles.Rivi}>
                                <label htmlFor={'nimiFi'}>{i18n.translate('LABEL_SUOMEKSI')}</label>
                                <div className={styles.PitkaInput}>
                                    <Input
                                        disabled={isDisabled}
                                        error={!!validationErrors['nimiFi']}
                                        id={'nimiFi'}
                                        {...register('nimiFi')}
                                        defaultValue={ryhma.nimi['fi']}
                                    />
                                </div>
                            </div>
                            <div className={styles.Rivi}>
                                <label htmlFor={'nimiSv'}>{i18n.translate('LABEL_RUOTSIKSI')}</label>
                                <div className={styles.PitkaInput}>
                                    <Input
                                        disabled={isDisabled}
                                        error={!!validationErrors['nimiSv']}
                                        id={'nimiSv'}
                                        {...register('nimiSv')}
                                        defaultValue={ryhma.nimi['sv']}
                                    />
                                </div>
                            </div>
                            <div className={styles.Rivi}>
                                <label htmlFor={'nimiEn'}>{i18n.translate('LABEL_ENGLANNIKSI')}</label>
                                <div className={styles.PitkaInput}>
                                    <Input
                                        disabled={isDisabled}
                                        error={!!validationErrors['nimiEn']}
                                        id={'nimiEn'}
                                        {...register('nimiEn')}
                                        defaultValue={ryhma.nimi['en']}
                                    />
                                </div>
                            </div>
                        </div>
                    </div>
                    <div className={styles.Rivi}>
                        <div className={styles.Kentta}>
                            <label>{i18n.translate('RYHMAT_RYHMAN_KUVAUS')}</label>
                            <div className={styles.Rivi}>
                                <label htmlFor={'kuvaus2Fi'}>{i18n.translate('LABEL_SUOMEKSI')}</label>
                                <div className={styles.PitkaInput}>
                                    <Input
                                        error={!!validationErrors['kuvaus2Fi']}
                                        id={'kuvaus2Fi'}
                                        {...register('kuvaus2Fi')}
                                        defaultValue={ryhma.kuvaus2['kieli_fi#1']}
                                    />
                                </div>
                            </div>
                            <div className={styles.Rivi}>
                                <label htmlFor={'kuvaus2Sv'}>{i18n.translate('LABEL_RUOTSIKSI')}</label>
                                <div className={styles.PitkaInput}>
                                    <Input
                                        error={!!validationErrors['kuvaus2Sv']}
                                        id={'kuvaus2Sv'}
                                        {...register('kuvaus2Sv')}
                                        defaultValue={ryhma.kuvaus2['kieli_sv#1']}
                                    />
                                </div>
                            </div>
                            <div className={styles.Rivi}>
                                <label htmlFor={'kuvaus2En'}>{i18n.translate('LABEL_ENGLANNIKSI')}</label>
                                <div className={styles.PitkaInput}>
                                    <Input
                                        error={!!validationErrors['kuvaus2En']}
                                        id={'kuvaus2En'}
                                        {...register('kuvaus2En')}
                                        defaultValue={ryhma.kuvaus2['kieli_en#1']}
                                    />
                                </div>
                            </div>
                        </div>
                    </div>
                    <div className={styles.Rivi}>
                        <div className={styles.Kentta}>
                            <label>{i18n.translate('RYHMAT_RYHMAN_TYYPPI')}</label>
                            <Controller
                                control={control}
                                name={'ryhmatyypit'}
                                defaultValue={ryhmaTyypit}
                                render={({ field }) => (
                                    <Select
                                        id="RYHMALOMAKE_RYHMAN_TYYPPI_SELECT"
                                        {...field}
                                        error={!!validationErrors['ryhmatyypit']}
                                        isMulti
                                        options={ryhmaTyypitOptions}
                                        isDisabled={isDisabled}
                                    />
                                )}
                            />
                        </div>
                    </div>
                    <div className={styles.Rivi}>
                        <div className={styles.Kentta}>
                            <label>{i18n.translate('RYHMAT_RYHMAN_KAYTTOTARKOITUS')}</label>
                            <Controller
                                control={control}
                                name={'kayttoryhmat'}
                                defaultValue={kayttoRyhmat}
                                render={({ field }) => (
                                    <Select
                                        id="RYHMALOMAKE_RYHMAN_KAYTTOTARKOITUS_SELECT"
                                        {...field}
                                        error={!!validationErrors['kayttoryhmat']}
                                        isMulti
                                        options={kayttoRyhmatOptions}
                                        isDisabled={isDisabled}
                                    />
                                )}
                            />
                        </div>
                    </div>
                    {!onUusi && (
                        <div className={styles.AlinRivi}>
                            <Button name="passivoibutton" variant="outlined" onClick={handlePassivoi}>
                                {ryhma.status === 'AKTIIVINEN'
                                    ? i18n.translate('RYHMAT_PASSIVOI_RYHMA')
                                    : i18n.translate('RYHMAT_AKTIVOI_RYHMA')}
                            </Button>
                            ,
                            <Button name="poistabutton" variant="outlined" onClick={handlePoista}>
                                {i18n.translate('RYHMAT_POISTA_RYHMA')}
                            </Button>
                        </div>
                    )}
                </div>
            </div>
            <div className={styles.AlaBanneri}>
                <div>
                    <Button
                        name="peruutabutton"
                        variant="outlined"
                        className={styles.Versionappula}
                        onClick={handlePeruuta}
                    >
                        {i18n.translate('BUTTON_SULJE')}
                    </Button>
                    <Button
                        disabled={ryhma.status === 'PASSIIVINEN'}
                        name="tallennabutton"
                        className={styles.Versionappula}
                        onClick={handleSubmit(handleTallenna)}
                    >
                        {i18n.translate('BUTTON_TALLENNA')}
                    </Button>
                </div>
            </div>
        </PohjaSivu>
    );
};

export default MuokkausLomake;
