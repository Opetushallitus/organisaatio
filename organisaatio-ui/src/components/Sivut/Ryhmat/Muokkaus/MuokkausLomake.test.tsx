import React from 'react';
import { shallow } from 'enzyme';
import '@testing-library/jest-dom/extend-expect';

import MuokkausLomake, { MuokkausLomakeProps } from './MuokkausLomake';

const MINIMAL_PROPS: Partial<MuokkausLomakeProps> = {
    nimiFiBind: {
        name: 'nimiFi',
        value: '1',
        onChange: jest.fn(),
    },
    nimiSvBind: {
        name: 'nimiSv',
        value: '2',
        onChange: jest.fn(),
    },
    nimiEnBind: {
        name: 'nimiEn',
        value: '3',
        onChange: jest.fn(),
    },
    kuvaus2FiBind: {
        name: 'kuvaus2Fi',
        value: '4',
        onChange: jest.fn(),
    },
    kuvaus2SvBind: {
        name: 'kuvaus2Sv',
        value: '5',
        onChange: jest.fn(),
    },
    kuvaus2EnBind: {
        name: 'kuvaus2En',
        value: '6',
        onChange: jest.fn(),
    },
    handleRyhmaSelectOnChange: jest.fn(),
    handlePeruuta: jest.fn(),
    handlePassivoi: jest.fn(),
    handlePoista: jest.fn(),
    handleTallenna: jest.fn(),
};

let testProps = { ...MINIMAL_PROPS };

afterAll(() => {
    jest.clearAllMocks();
});

beforeEach(() => {
    testProps = { ...MINIMAL_PROPS };
    jest.resetAllMocks();
});

describe('MuokkausLomake', () => {
    describe('Rendering', () => {
        it('Renders without crashing', () => {
            const wrapper = shallow(<MuokkausLomake {...testProps as MuokkausLomakeProps} />);
            expect(wrapper).toMatchSnapshot();
        });
    });
    describe('Update nimi in all languages', () => {
        test.each([['nimiFi'], ['nimiSv'], ['nimiEn']])('%s', (name) => {
            const wrapper = shallow(<MuokkausLomake {...testProps as MuokkausLomakeProps} />);
            const field = wrapper.find({ name });
            field.simulate('change', { target: { name, value: 'testiNimi' } });
            expect(testProps[`${name}Bind`].onChange).toHaveBeenCalledTimes(1);
            expect((testProps[`${name}Bind`].onChange as jest.Mock).mock.calls).toMatchSnapshot();
        });
    });

    describe('Update kuvaus2 in all languages', () => {
        test.each([['kuvaus2Fi'], ['kuvaus2Sv'], ['kuvaus2En']])('%s', (name) => {
            const wrapper = shallow(<MuokkausLomake {...testProps as MuokkausLomakeProps} />);
            const field = wrapper.find({ name });
            field.simulate('change', { target: { name, value: 'testiKuvaus2' } });
            expect(testProps[`${name}Bind`].onChange).toHaveBeenCalledTimes(1);
            expect((testProps[`${name}Bind`].onChange as jest.Mock).mock.calls).toMatchSnapshot();
        });
    });

    describe('Buttons', () => {
        describe('Passivoi button', () => {
            it('clicks', () => {
                const wrapper = shallow(<MuokkausLomake {...testProps as MuokkausLomakeProps} />);
                const button = wrapper.find({ name: 'passivoibutton' });
                button.simulate('click');
                expect(testProps.handlePassivoi).toHaveBeenCalledTimes(1);
                expect((testProps.handlePassivoi as jest.Mock).mock.calls).toMatchSnapshot();
            });
        });
        describe('Poista button', () => {
            it('clicks', () => {
                const wrapper = shallow(<MuokkausLomake {...testProps as MuokkausLomakeProps} />);
                const button = wrapper.find({ name: 'poistabutton' });
                button.simulate('click');
                expect(testProps.handlePoista).toHaveBeenCalledTimes(1);
                expect((testProps.handlePoista as jest.Mock).mock.calls).toMatchSnapshot();
            });
        });

        describe('Tallenna button', () => {
            it('clicks', () => {
                const wrapper = shallow(<MuokkausLomake {...testProps as MuokkausLomakeProps} />);
                const button = wrapper.find({ name: 'tallennabutton' });
                button.simulate('click');
                expect(testProps.handleTallenna).toHaveBeenCalledTimes(1);
                expect((testProps.handleTallenna as jest.Mock).mock.calls).toMatchSnapshot();
            });
        });

        describe('Peruuta button', () => {
            it('clicks', () => {
                const wrapper = shallow(<MuokkausLomake {...testProps as MuokkausLomakeProps} />);
                const button = wrapper.find({ name: 'peruutabutton' });
                button.simulate('click');
                expect(testProps.handlePeruuta).toHaveBeenCalledTimes(1);
                expect((testProps.handlePeruuta as jest.Mock).mock.calls).toMatchSnapshot();
            });
        });
    });
});
