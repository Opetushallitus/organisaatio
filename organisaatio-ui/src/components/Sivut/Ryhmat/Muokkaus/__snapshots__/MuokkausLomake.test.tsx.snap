// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Array [
    Object {
      "target": Object {
        "name": "nimiFi",
        "value": "testiNimi",
      },
    },
  ],
]
`;

exports[` 2`] = `
Array [
  Array [
    Object {
      "target": Object {
        "name": "nimiSv",
        "value": "testiNimi",
      },
    },
  ],
]
`;

exports[` 3`] = `
Array [
  Array [
    Object {
      "target": Object {
        "name": "nimiEn",
        "value": "testiNimi",
      },
    },
  ],
]
`;

exports[` 4`] = `
Array [
  Array [
    Object {
      "target": Object {
        "name": "kuvaus2Fi",
        "value": "testiKuvaus2",
      },
    },
  ],
]
`;

exports[` 5`] = `
Array [
  Array [
    Object {
      "target": Object {
        "name": "kuvaus2Sv",
        "value": "testiKuvaus2",
      },
    },
  ],
]
`;

exports[` 6`] = `
Array [
  Array [
    Object {
      "target": Object {
        "name": "kuvaus2En",
        "value": "testiKuvaus2",
      },
    },
  ],
]
`;

exports[`MuokkausLomake Rendering Renders without crashing 1`] = `
<PohjaSivu>
  <div
    className="YlaBanneri"
  >
    <div>
      <a
        href="/organisaatio/ryhmat"
      >
        <Icon
          icon={
            Object {
              "body": "<path d=\\"M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z\\" fill=\\"currentColor\\"/>",
              "height": 512,
              "width": 576,
            }
          }
        />
      </a>
    </div>
    <div>
      <a
        href="/organisaatio/ryhmat"
      >
        KAIKKI_RYHMAT
      </a>
    </div>
  </div>
  <div
    className="PaaKehys"
  >
    <div
      className="ValiContainer"
    >
      <div
        className="ValiOtsikko"
      >
        <h3>
          RYHMA
        </h3>
        <h1>
          Suominimi
        </h1>
      </div>
    </div>
    <div
      className="PaaOsio"
    >
      <div
        className="OtsikkoRivi"
      >
        <div
          className="Otsikko"
        >
          <h3>
            RYHMAN_TIEDOT_OTSIKKO
          </h3>
        </div>
      </div>
      <div
        className="OidRivi"
      >
        <span
          className="AvainKevyestiBoldattu"
        >
          OID
        </span>
        <span
          className="ReadOnly"
        >
          1234
        </span>
      </div>
      <div
        className="Rivi"
      >
        <div
          className="Kentta"
        >
          <label>
            RYHMAN_NIMI
          </label>
          <InputRivi
            bind={
              Object {
                "localizationKey": "SUOMEKSI",
                "name": "nimiFi",
                "onChange": [MockFunction],
                "value": "1",
              }
            }
            key="nimiFi"
          />
          <InputRivi
            bind={
              Object {
                "localizationKey": "RUOTSIKSI",
                "name": "nimiSv",
                "onChange": [MockFunction],
                "value": "2",
              }
            }
            key="nimiSv"
          />
          <InputRivi
            bind={
              Object {
                "localizationKey": "ENGLANNIKSI",
                "name": "nimiEn",
                "onChange": [MockFunction],
                "value": "3",
              }
            }
            key="nimiEn"
          />
        </div>
      </div>
      <div
        className="Rivi"
      >
        <div
          className="Kentta"
        >
          <label>
            RYHMAN_KUVAUS
          </label>
          <InputRivi
            bind={
              Object {
                "localizationKey": "SUOMEKSI",
                "name": "kuvaus2Fi",
                "onChange": [MockFunction],
                "value": "4",
              }
            }
            key="kuvaus2Fi"
          />
          <InputRivi
            bind={
              Object {
                "localizationKey": "RUOTSIKSI",
                "name": "kuvaus2Sv",
                "onChange": [MockFunction],
                "value": "5",
              }
            }
            key="kuvaus2Sv"
          />
          <InputRivi
            bind={
              Object {
                "localizationKey": "ENGLANNIKSI",
                "name": "kuvaus2En",
                "onChange": [MockFunction],
                "value": "6",
              }
            }
            key="kuvaus2En"
          />
        </div>
      </div>
      <div
        className="Rivi"
      >
        <div
          className="Kentta"
        >
          <label>
            RYHMAN_TYYPPI
          </label>
          <Select
            isDisabled={false}
            isMulti={true}
            name="ryhmatyypit"
            onChange={[MockFunction]}
            options={Array []}
            value={Array []}
          />
        </div>
      </div>
      <div
        className="Rivi"
      >
        <div
          className="Kentta"
        >
          <label>
            RYHMAN_KAYTTOTARKOITUS
          </label>
          <Select
            isDisabled={false}
            isMulti={true}
            name="kayttoryhmat"
            onChange={[MockFunction]}
            options={Array []}
            value={Array []}
          />
        </div>
      </div>
      <div
        className="AlinRivi"
      >
        <ForwardRef
          name="passivoibutton"
          onClick={[MockFunction]}
          variant="outlined"
        >
          PASSIVOI_RYHMA
        </ForwardRef>
        ,
        <ForwardRef
          name="poistabutton"
          onClick={[MockFunction]}
          variant="outlined"
        >
          POISTA_RYHMA
        </ForwardRef>
      </div>
    </div>
  </div>
  <div
    className="AlaBanneri"
  >
    <div>
      <ForwardRef
        className="Versionappula"
        name="peruutabutton"
        onClick={[MockFunction]}
        variant="outlined"
      >
        SULJE_TIEDOT
      </ForwardRef>
      <ForwardRef
        className="Versionappula"
        disabled={false}
        name="tallennabutton"
        onClick={[MockFunction]}
      >
        TALLENNA
      </ForwardRef>
    </div>
  </div>
</PohjaSivu>
`;
